File: OpenCount/Services/SampleSessionSeeder.swift
import SwiftUI

// MARK: - SampleSessionSeeder

/// Seeds the bundled `SampleSession.json` fixture into StorageService on first launch.
///
/// Requirements: 29.4, 29.5
enum SampleSessionSeeder {

    private static let seededKey = "hasSeedSampleSession"

    /// Seeds the sample session if it has not been seeded before.
    /// - Parameter force: When true, seeds even if previously created.
    @MainActor
    static func seedIfNeeded(force: Bool = false) async {
        guard force || !UserDefaults.standard.bool(forKey: seededKey) else { return }
        guard let session = loadFixture() else { return }
        try? await StorageService.shared.save(session)
        UserDefaults.standard.set(true, forKey: seededKey)
    }

    /// Seeds into an AppState (called from AppState.seedSampleIfNeeded).
    @MainActor
    static func seed(into appState: AppState) {
        guard let session = loadFixture() else { return }
        appState.sessions.append(session)
        Task { try? await StorageService.shared.save(session) }
    }

    // MARK: - Fixture loading

    private static func loadFixture() -> CountSession? {
        guard let url = Bundle.main.url(forResource: "SampleSession", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let dto = try? JSONDecoder().decode(SampleSessionDTO.self, from: data)
        else {
            return nil
        }

        let session = CountSession(
            name: dto.name,
            sessionDescription: dto.sessionDescription
        )

        var objectTypeMap: [String: ObjectType] = [:]
        for otDTO in dto.objectTypes {
            let ot = ObjectType(
                name: otDTO.name,
                colorHex: otDTO.colorHex,
                iconName: otDTO.iconName,
                sortOrder: otDTO.sortOrder,
                session: session
            )
            objectTypeMap[otDTO.id] = ot
        }
        session.objectTypes = Array(objectTypeMap.values).sorted { $0.sortOrder < $1.sortOrder }

        var markers: [CountMarker] = []
        for mDTO in dto.markers {
            guard let objectType = objectTypeMap[mDTO.objectTypeID] else { continue }
            let marker = CountMarker(
                normalizedX: mDTO.normalizedX,
                normalizedY: mDTO.normalizedY,
                objectType: objectType,
                isAIDerived: mDTO.isAIDerived,
                session: session
            )
            markers.append(marker)
        }
        session.markers = markers

        var regions: [CountRegion] = []
        for rDTO in dto.regions {
            let points = rDTO.normalizedPoints.map { CGPoint(x: $0.x, y: $0.y) }
            let shapeType = RegionShapeType(rawValue: rDTO.shapeType) ?? .rectangle
            let region = CountRegion(
                name: rDTO.name,
                colorHex: rDTO.colorHex,
                shapeType: shapeType,
                normalizedPoints: points,
                session: session
            )
            regions.append(region)
        }
        session.regions = regions

        return session
    }
}

// MARK: - DTO types

private struct SampleSessionDTO: Decodable {
    let id: String
    let name: String
    let sessionDescription: String?
    let objectTypes: [ObjectTypeDTO]
    let markers: [MarkerDTO]
    let regions: [RegionDTO]
}

private struct ObjectTypeDTO: Codable {
    let id: String
    let name: String
    let colorHex: String
    let iconName: String
    let sortOrder: Int
}

private struct MarkerDTO: Codable {
    let id: String
    let normalizedX: Double
    let normalizedY: Double
    let objectTypeID: String
    let isAIDerived: Bool
}

private struct RegionDTO: Codable {
    let id: String
    let name: String
    let colorHex: String
    let shapeType: String
    let normalizedPoints: [PointDTO]
}

private struct PointDTO: Codable {
    let x: Double
    let y: Double
}

File: packages/shared/src/commonMain/kotlin/com/opencount/shared/service/SampleSessionSeeder.kt
package com.opencount.shared.service

import com.opencount.shared.model.CountSession
import com.opencount.shared.model.CountMarker
import com.opencount.shared.model.ObjectType
import com.opencount.shared.model.NormalizedPoint
import com.opencount.shared.model.CountRegion
import com.opencount.shared.model.RegionShapeType
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

expect fun persistFlag(key: String, value: Boolean)
expect fun readFlag(key: String): Boolean

object SampleSessionSeeder {
    private const val SEEN_KEY = "sample_session_seeded"

    fun seedIfNeeded(storage: StorageService, force: Boolean = false) {
        if (!force && readFlag(SEEN_KEY)) return
        val session = createSampleSession()
        storage.save(session)
        persistFlag(SEEN_KEY, true)
    }

    fun createSampleSession(): CountSession {
        val types = listOf(
            ObjectType.create("Cars", colorHex = "#FF5733"),
            ObjectType.create("People", colorHex = "#33FF57"),
            ObjectType.create("Trees", colorHex = "#3357FF"),
        )
        val markers = listOf(
            CountMarker.create(0.1, 0.2, types[0].id),
            CountMarker.create(0.3, 0.4, types[0].id),
            CountMarker.create(0.5, 0.6, types[0].id),
            CountMarker.create(0.2, 0.3, types[1].id),
            CountMarker.create(0.4, 0.5, types[1].id),
            CountMarker.create(0.6, 0.7, types[2].id),
        )
        val regions = listOf(
            CountRegion(
                id = "sample-region-1",
                name = "Parking Lot",
                shapeType = RegionShapeType.Rectangle,
                normalizedPoints = listOf(NormalizedPoint(0.0, 0.0), NormalizedPoint(0.5, 0.5)),
            ),
        )
        val zonedMarkers = markers.mapIndexed { i, m ->
            if (i < 4) m.copy(regionId = "sample-region-1") else m
        }

        return CountSession.create("Sample Count").copy(
            objectTypes = types,
            markers = zonedMarkers,
            regions = regions,
            sessionDescription = "Welcome to OpenCount! This sample session shows how counting works.",
        )
    }

    fun sampleJSON(): String {
        val json = Json { prettyPrint = true }
        return json.encodeToString(CountSession.serializer(), createSampleSession())
    }
}
